controller/volume_rebuilding_controller.go

```diff
@@ -190,14 +252,124 @@ func (vbc *VolumeRebuildingController) reconcile(volName string) (err error) {
 			return
 		}
 	}()
+	existingVol := vol.DeepCopy()
+	defer func() {
+		if err != nil {
+			return
+		}
+		if reflect.DeepEqual(existingVol.Spec, vol.Spec) {
+			return
+		}
+
+		if _, err = vbc.ds.UpdateVolume(vol); err != nil {
+			return
+		}
+	}()
 
 	rebuildingAttachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeVolumeRebuildingController, volName)
+	csiTicketExists := isVATicketExists(va, longhorn.AttacherTypeCSIAttacher)
+	isOfflineRebuildEnabled, err := vbc.isVolumeOfflineRebuildEnabled(vol.Spec.OfflineRebuild)
+	if err != nil {
+		return err
+	}
 
-	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+	if isOfflineRebuildEnabled && vol.Status.Robustness != longhorn.VolumeRobustnessFaulted && !csiTicketExists {
+		if vol.Status.State == longhorn.VolumeStateDetached {
+			va, err = vbc.syncLHVolumeAttachementForOfflineRebuild(vol, va, rebuildingAttachmentTicketID)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		if vol.Status.State == longhorn.VolumeStateAttaching {
+			return nil
+		}
+
+		isVolumeInRebuilding, err := vbc.isVolumeReplicasRebuilding(vol)
+		if err != nil {
+			return err
+		}
+		if isVolumeInRebuilding {
+			return nil
+		}
+	}
+
+	if isOfflineRebuildEnabled && vol.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+		warnMsg := fmt.Sprintf("Volume %v is faulted, skip rebuilding", volName)
+		vbc.logger.Info(warnMsg)
+		vbc.eventRecorder.Event(vol, corev1.EventTypeWarning, constant.EventReasonCanceledOfflineRebuild, warnMsg)
+	}
 
+	if !isOfflineRebuildEnabled && isVATicketExists(va, longhorn.AttacherTypeVolumeRebuildingController) {
+		warnMsg := fmt.Sprintf("Canceling volume %v offline rebuilding", volName)
+		vbc.logger.Info(warnMsg)
+		vbc.eventRecorder.Event(vol, corev1.EventTypeWarning, constant.EventReasonCanceledOfflineRebuild, warnMsg)
+	}
+
+	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+	vol.Spec.DisableFrontend = false
 	return nil
 }
 
+func (vbc *VolumeRebuildingController) isVolumeOfflineRebuildEnabled(offlineRebuild longhorn.VolumeOfflineRebuild) (bool, error) {
+	offlineReplicaRebuildingValue, err := vbc.ds.GetSettingAsBool(types.SettingNameOfflineReplicaRebuilding)
+	if err != nil {
+		return false, err
+	}
+	return (offlineRebuild == longhorn.VolumeOfflineRebuildEnabled) || (offlineRebuild == longhorn.VolumeOfflineRebuildIgnored && offlineReplicaRebuildingValue), nil
+}
+
+func (vbc *VolumeRebuildingController) syncLHVolumeAttachementForOfflineRebuild(vol *longhorn.Volume, va *longhorn.VolumeAttachment, attachmentID string) (*longhorn.VolumeAttachment, error) {
+	replicas, err := vbc.ds.ListVolumeReplicasRO(vol.Name)
+	if err != nil {
+		return va, err
+	}
+	if !util.IsVolumeReplicasHealthy(vol.Spec.NumberOfReplicas, replicas) {
+		createOrUpdateAttachmentTicket(va, attachmentID, vol.Status.OwnerID, longhorn.TrueValue, longhorn.AttacherTypeVolumeRebuildingController)
+		vol.Spec.DisableFrontend = true
```

Comment:
Why do we need to disable frontend?

Reply:
Let it get into the `maintenence` mode.

Reply:
I see. 
Now thinking about it again, I think we can do something similar to https://github.com/longhorn/longhorn-manager/blob/763ac9117eb243349d7d91827a1023eaa7980f35/controller/snapshot_controller.go#L563-L564 by setting `longhorn.AttachmentParameterDisableFrontend` to  `longhorn.AnyValue` instead of hardcode the `vol.Spec.DisableFrontend = true`

I think for rebuilding, front-end does not matter. 

WDYT?

Reply:
>by setting longhorn.AttachmentParameterDisableFrontend to longhorn.AnyValue instead of hardcode the vol.Spec.DisableFrontend = true

Do you mean to set `longhorn.AttachmentParameterDisableFrontend` to `longhorn.TrueValue` because we want volumes in the `maintenance` mode?

Reply:
> Do you mean to set longhorn.AttachmentParameterDisableFrontend to longhorn.TrueValue because we want volumes in the maintenance mode?

I mean to set longhorn.AttachmentParameterDisableFrontend to longhorn.TrueValue because we don't care whether the volume has frontenddisable or not. In either case, the rebuilding operation can proceed

Reply:
Modified. Thanks.